File: problem/Minimum_Score/minimumScore.py
from collections import defaultdict, deque

class Solution:
    def minimumScore(self, nums, edges):
        n = len(nums)
        m = len(edges)
        
        graph = defaultdict(list)
        children = defaultdict(set)
        arr = nums[:]
        degree = [0]*n
        
        for x, y in edges:
            graph[x].append(y)
            graph[y].append(x)
            degree[x] += 1
            degree[y] += 1
        
        v = 0
        seen = set()
        q = deque()
        for i in range(n):
            v ^= nums[i]
            if degree[i] == 1:
                q.append(i)
                seen.add(i)
        
        while q:
            curr = q.popleft()
            for nxt in graph[curr]:
                if nxt not in seen:
                    children[nxt].add(curr)
                    children[nxt] |= children[curr]
                    arr[nxt] ^= arr[curr]
                degree[nxt] -= 1
                if degree[nxt] == 1:
                    seen.add(nxt)
                    q.append(nxt)
        
        ans = float('inf')
        for i in range(m-1):
            for j in range(i+1, m):
                a, b = edges[i]
                if b in children[a]:
                    a, b = b, a
                c, d = edges[j]
                if d in children[c]:
                    c, d = d, c
                
                if c in children[a]:
                    curr = [arr[c], arr[c]^arr[a], v^arr[a]]
                elif a in children[c]:
                    curr = [arr[a], arr[c]^arr[a], v^arr[c]]
                else:
                    curr = [arr[a], arr[c], v^arr[a]^arr[c]]
                ans = min(ans, max(curr)-min(curr))
        return ans

File: problem/Minimum_Score/minimumScore.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAXN 1005
#define INF 1000000000

int nums[MAXN];
int child_xor[MAXN];
int childs[MAXN][MAXN];
int visited[MAXN];
int adj[MAXN][MAXN], adjSize[MAXN];
int pc[MAXN * 2][2], pcSize = 0;
int par[MAXN], parSize = 0;

int max3(int a, int b, int c)
{
    int max = a > b ? a : b;
    return max > c ? max : c;
}

int min3(int a, int b, int c)
{
    int min = a < b ? a : b;
    return min < c ? min : c;
}

int dfs(int node)
{
    visited[node] = 1;
    int ans = nums[node];

    for (int i = 0; i < parSize; ++i)
        childs[par[i]][node] = 1;

    par[parSize++] = node;

    for (int i = 0; i < adjSize[node]; ++i)
    {
        int child = adj[node][i];
        if (!visited[child])
        {
            pc[pcSize][0] = node;
            pc[pcSize][1] = child;
            pcSize++;
            ans ^= dfs(child);
        }
    }

    parSize--;
    child_xor[node] = ans;
    return ans;
}

int minimumScore(int *numsInput, int numsSize, int **edges, int edgesSize, int *edgesColSize)
{
    for (int i = 0; i < numsSize; ++i)
        nums[i] = numsInput[i];

    pcSize = 0;
    parSize = 0;
    memset(adjSize, 0, sizeof(adjSize));
    memset(visited, 0, sizeof(visited));
    memset(childs, 0, sizeof(childs));
    memset(child_xor, 0, sizeof(child_xor));

    for (int i = 0; i < edgesSize; ++i)
    {
        int u = edges[i][0];
        int v = edges[i][1];
        adj[u][adjSize[u]++] = v;
        adj[v][adjSize[v]++] = u;
    }

    dfs(0);

    int ans = INF;

    for (int i = 0; i < pcSize; ++i)
    {
        for (int j = i + 1; j < pcSize; ++j)
        {
            int a = pc[i][1];
            int b = pc[j][1];

            int xa = child_xor[a];
            int xb = child_xor[b];
            int xc = child_xor[0];

            if (childs[a][b])
            {
                xc ^= xa;
                xa ^= xb;
            }
            else if (childs[b][a])
            {
                xc ^= xb;
                xb ^= xa;
            }
            else
            {
                xc ^= xa;
                xc ^= xb;
            }

            int mx = max3(xa, xb, xc);
            int mn = min3(xa, xb, xc);
            if (ans > mx - mn)
                ans = mx - mn;
        }
    }

    return ans;
}